onnx/defs/nn/defs.cc

```diff
@@ -1534,37 +1534,49 @@ ONNX_OPERATOR_SET_SCHEMA(
             "var",
             "running (training) or estimated (testing) variance tensor of shape (C).",
             "T")
+        .Input(
+            5,
+            "training_mode",
+            "If set to true, run spatial batch normalization in training mode, default is false.",
+            "T1",
+            OpSchema::Optional)
         .Output(0, "Y", "The output tensor of the same shape as X", "T")
         .Output(
             1,
             "mean",
-            "The running mean after the BatchNormalization operator.",
+            "The running mean after the BatchNormalization operator."
+            "This output is populated only when training_mode is set to true.",
             "T",
```

Comment:
Those outputs (mean, var, saved_mean, saved_var) cannot be consumed by any other operators. Otherwise, SSA assumption will be broken. Could we make it explicitly here by saying that `this output cannot be an input of any other operator`.

Reply:
makes sense; I will add it explicitly in the doc.

Reply:
Let's just make `mean` and `var` inaccessible. The other two will be accessed by training operators and they don't break SSA.

Reply:
I think we just need to (locally) rename the outputs of the operator to be different from the inputs of the operator for the sake of clarity: e.g., use, output_mean and output_var instead of mean and var. I don't think there needs to be any constraints like "this output cannot be an input of any other operator".

Reply:
The names used here are "formal parameter names", and have no connection to the "actual parameter names" used in an actual node containing a call to this op.